src/components/CheckboxGroup/CheckboxGroup.tsx

```diff
@@ -0,0 +1,133 @@
+import React, { useEffect, useId, useReducer } from 'react';
+import cn from 'classnames';
+
+import { Checkbox, ErrorMessage } from '@/components';
+
+import classes from './CheckboxGroup.module.css';
+
+export interface CheckboxItem {
+  checked?: boolean;
+  description?: string;
+  disabled?: boolean;
+  id?: string;
+  label?: string;
+  name: string;
+}
+
+export enum CheckboxGroupVariant {
+  Vertical = 'vertical',
+  Horisontal = 'horisontal',
+}
+
+export type CheckedNames = string[];
+
+export interface CheckboxGroupProps {
+  compact?: boolean;
+  description?: string;
+  disabled?: boolean;
+  error?: string;
+  items: CheckboxItem[];
+  legend?: string;
+  onChange?: (names: CheckedNames) => void;
+  variant?: CheckboxGroupVariant;
+}
+
+type ReducerAction =
+  | { type: 'check' | 'uncheck'; name: string }
+  | { type: 'reset'; state: CheckedNames };
+
+const reducer = (state: CheckedNames, action: ReducerAction) => {
+  switch (action.type) {
+    case 'check':
+      return state.concat([action.name]);
+    case 'uncheck':
+      return state.filter((name) => name !== action.name);
+    case 'reset':
+      return action.state;
+  }
+};
+const checkedItems = (items: CheckboxItem[]) =>
+  items.filter(({ checked }) => checked).map(({ name }) => name);
+
+export const CheckboxGroup = ({
+  compact,
+  description,
+  disabled,
+  error,
+  items,
+  legend,
+  onChange,
+  variant = CheckboxGroupVariant.Vertical,
+}: CheckboxGroupProps) => {
+  const allNames = items.map((item) => item.name);
+  if (allNames.length !== new Set(allNames).size) {
+    throw Error('Each name in the checkbox group must be unique.');
+  }
+
+  const randomId = useId();
+
+  const [checkedNames, dispatch] = useReducer(reducer, checkedItems(items));
+
+  useEffect(
+    () => dispatch({ type: 'reset', state: checkedItems(items) }),
+    [items],
+  );
+
+  useEffect(
+    () => (onChange && !disabled ? onChange(checkedNames) : undefined),
+    [checkedNames, onChange, disabled],
+  );
+
+  return (
+    <div
+      className={cn(
+        classes.wrapper,
+        error && classes['wrapper--error'],
+        compact && classes['wrapper--compact'],
+        disabled && classes['wrapper--disabled'],
+      )}
+    >
+      {error && <div className={classes['wrapper--error__line']} />}
+      <fieldset className={classes['checkbox-group']}>
+        {legend && <legend>{legend}</legend>}
+        {description && (
+          <p className={classes['checkbox-group__description']}>
+            {description}
+          </p>
+        )}
+        <div
+          className={cn(
+            classes['checkbox-group__list'],
+            classes[`checkbox-group__list--${variant}`],
+          )}
+        >
+          {items.map((item, index) => (
+            <div key={`checkbox-group-${randomId}-${index}`}>
```

Comment:
[Using `index` as keys should be avoided](https://robinpokorny.com/blog/index-as-a-key-is-an-anti-pattern/). Adding `randomId` to it wont help here, because that will be the same for all items in the group, effectively making the only difference between items the `index`.

I would suggest to just use `name` as the identifier here, then you can also drop the `randomId`. The `name` is required, and also required to be unique. 

The key can also be directly attached to the `Checkbox` if you want to avoid the extra div.

```suggestion
            <div key={item.name}>
```

Reply:
The `randomId` was just to distinguish the items from other elements used by the consumer. Anyways, I wasn't aware about this, very useful to know 🙂

Reply:
Theres no need to distinguish these from a consumers point of view. The `key` is an internal React thing only. Its also fine to use identical keys in different "instances" because they will not interfere with each other.